ignition-server/controllers/image_file_cache.go

```diff
@@ -0,0 +1,145 @@
+package controllers
+
+import (
+	"bytes"
+	"context"
+	"crypto/sha256"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"sync"
+
+	"github.com/openshift/hypershift/support/releaseinfo/registryclient"
+	ctrl "sigs.k8s.io/controller-runtime"
+)
+
+var log = ctrl.Log.WithName("image-cache")
+
+func NewImageFileCache(workDir string) (*imageFileCache, error) {
+	cacheDir, err := os.MkdirTemp(workDir, "cache-*")
+	if err != nil {
+		return nil, fmt.Errorf("failed to create cache directory: %w", err)
+	}
+
+	return &imageFileCache{
+		cacheMap:  make(map[cacheKey]cacheValue),
+		cacheDir:  cacheDir,
+		regClient: registryclient.ExtractImageFile,
+		RWMutex:   sync.RWMutex{},
+	}, nil
+}
+
+// Caching wrapper around the registry client function "regClient".
+// All files returned by the function are cached as files in a sub-directory of "cacheDir".
+// The cached files never expire, but can be automatically re-downloaded if they become corrupted.
+type imageFileCache struct {
+	cacheMap  map[cacheKey]cacheValue
+	cacheDir  string
+	regClient regClient // wrapped registryclient function
+	sync.RWMutex
+}
+
+type checksum []byte
+type regClient func(ctx context.Context, imageRef string, pullSecret []byte, file string, out io.Writer) error
+
+type cacheKey struct {
+	imageRef  string
+	imageFile string
+}
+
+type cacheValue struct {
+	fileName string
+	fileSha  checksum
+}
+
+func (c *imageFileCache) extractImageFile(ctx context.Context, imageRef string, pullSecret []byte, imageFile string, out io.Writer) error {
+	c.RLock()
+	defer c.RUnlock()
+
+	key := cacheKey{imageRef: imageRef, imageFile: imageFile}
+
+	if file, ok := c.cacheMap[key]; ok && cacheFileIsValid(file) {
+		log.Info("retrieved cached file", "imageRef", imageRef, "file", imageFile)
+		return returnCacheFile(file.fileName, out)
+	}
+
+	// image file is not in the cache or checksum test failed - download from the registry...
```

Comment:
I believe there's a mutex race condition here: shouldn't at this point you release the read lock and claim the write lock?

Reply:
As tyler writes. You also need to consider that between trying to acquire the w lock and actually getting it, another routine could have fetched the entry, so you need to check again once you got the w lock.

Reply:
Yes, there was a race condition and now it is fixed. The whole function is one big critical section and a reader lock cannot be used. The only simple and safe optimization I could make is to unlock mutex before blocking on returning cached file in the happy case scenario.